### ISSUE TITLE
TransformerDocumentEmbeddings vs TransformerWordEmbeddings

### ISSUE THREAD

**Issue by Bunoviske** [flairNLP/flair#1761]
Hello! I was trying to figure out the implementation details of these embeddings, but I had difficulty. I understand that TransformerWordEmbedding return the embedding for each word and the TransformerDocumentEmbedding returns just one embedding for the whole document. My questions are:

1- How TransformerDocumentEmbedding pools the information of each word embedding? I couldn't understand very well the post processing after getting the transformer hidden layers:

            # put encoded batch through transformer model to get all hidden states of all encoder layers
            hidden_states = self.model(input_ids, attention_mask=mask)[-1] if len(sentences) > 1 \
                else self.model(input_ids)[-1]

            # iterate over all subtokenized sentences
            for sentence_idx, (sentence, subtokens) in enumerate(zip(sentences, subtokenized_sentences)):

                index_of_CLS_token = 0 if self.initial_cls_token else len(subtokens) -1

                cls_embeddings_all_layers: List[torch.FloatTensor] = \
                    [hidden_states[layer][sentence_idx][index_of_CLS_token] for layer in self.layer_indexes]

                # use scalar mix of embeddings if so selected
                if self.use_scalar_mix:
                    sm = ScalarMix(mixture_size=len(cls_embeddings_all_layers))
                    sm_embeddings = sm(cls_embeddings_all_layers)

                    cls_embeddings_all_layers = [sm_embeddings]

                # set the extracted embedding for the token
                sentence.set_embedding(self.name, torch.cat(cls_embeddings_all_layers)) 

2- Now TransformerWordEmbedding support the feature "allow_long_sentence". That problem can also happen on the TransformerDocumentEmbedding? Or its implementation already solve the issue of long sentences?

3- What is the main difference between these two embeddings before feeding the transformer model? Clearly there are a lot of different preprocessing before giving a Sentence to the transformer model.

Thanks for this great lib for the community!

**Comment by alanakbik** [comment]
Hello @Bunoviske the `TransformerWordEmbeddings` have to reconcile the output of diferent tokenizers, which is why so much pre-processing is happening. The transformers do subtokenization of words, so we need to find out which word was split into how many subtokens in order to get embeddings for each word. 

This is different in `TransformerDocumentEmbeddings`. Since we don't need word embeddings here, we don't have to do this. Each transformer has a special token from which the sentence-level embedding is extracted. So in both cases, the same transformer is used, but in the one we extract word embeddings and in the other sentence embeddings. 

`TransformerDocumentEmbeddings` currently truncate overlong text to the maximum number of subtokens. So long texts are handled differently here.

**Comment by Bunoviske** [comment]
I was not aware of the CLS token. Besides that, what do you recommend for texts bigger than 512 subtokens in TextClassification?

One though is implementing the same algorithm as here [TransformerWordEmbedding](https://github.com/flairNLP/flair/pull/1680) and pooling the CLS from all the sentences in the big text. I can't see a clear way to customize this use case on my own without manipulating flair code. Would this be a nice feature to add?

Thanks again :)
